### ISSUE TITLE
overlapping_range_endpoints false positive

### ISSUE THREAD

**Issue by benluelo** [rust-lang/rust#117648]
### Code

```Rust
match (0i8, 0i8) {
    (0, _) => {}
    // change either of these to not overlap and it will error
    (..=-1, ..=0) => {}
    (1.., 0..) => {}
    (1.., ..=-1) | (..=-1, 1..) => {}
}
```

### Current output

```Shell
warning: multiple patterns overlap on their endpoints
 --> src/main.rs:6:15
  |
5 |         (..=-1, ..=0) => {}
  |                 ---- this range overlaps on `0_i8`...
6 |         (1.., 0..) => {}
  |               ^^^ ... with this range
  |
  = note: you likely meant to write mutually exclusive ranges
  = note: `#[warn(overlapping_range_endpoints)]` on by default
```

### Desired output

```Shell
no output, pattern is exhaustive. addressing the warning causes an error as it is no longer exhaustive.
```

### Anything else?

this is only occurring on nightly (at least as of 2023-11-01, i haven't tested anything sooner)

**Comment by Nadrieril** [comment]
Yep, that's somewhat intentional. The true reason is that detecting this properly would require more work from exhaustiveness checking, which is a performance cost I predicted we wouldn't want to pay. That said I haven't actually measured anything so I could be wrong.

The rationalization is that forcing the user to clarify intent could be worth it for future readers. One way to clarify would be to write `..0 | 0` instead of `..=0`. I won't argue that this is particularly clearer in your case :sweat_smile:.

I won't close this as `wontfix` until we can decide based on perf measurements, but I also don't have the intention to do this work myself so this may not be fixed for a long while sorry.

**Comment by scooter-dangle** [comment]
This is showing up in a way that will cause confusion for devs in our codebase. Would it be possible to move this warning in its current form to a clippy lint given that it has easily confirmed false positives?

**Comment by Nadrieril** [comment]
It's unfortunately not possible to split off the FP-prone portion. The whole lint is FP-prone from the way it is currently computed :/

**Comment by Nadrieril** [comment]
This turned out to be pretty easy to implement! (https://github.com/rust-lang/rust/pull/119396) Let's see what the perf report says.

**Comment by Nadrieril** [comment]
The perf report reports a perf improvement x) I'm glad I tried, the PR should get merged

**Comment by Nadrieril** [comment]
Fixed by https://github.com/rust-lang/rust/pull/119396 :tada:
